### PR TITLE
Additional hooks for inventory manipulations callbacks.

### PR DESCRIPTION
**Pull request by Foghrye4**
With three new callbacks (on inventory itemstack add, move and drop) modders will be able to trace inventory state at any moment without need to iterate thru all players inventories. This way we can make, for example, overweight inventory effects.

### CODE REVIEW

File: src/inventorymanager.cpp [hunk 1]
```diff
@@ -523,6 +523,43 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
 		}
 	}
 
+	InventoryLocation from_loc;
+	if(from_inv.type == InventoryLocation::NODEMETA)
```
Comment (rubenwardy) [inline src/inventorymanager.cpp]: we use

```cpp
if (condition) {
    code;
}
```

Comment (Foghrye4) [inline src/inventorymanager.cpp]: Should i edit my patch immediately to this pull accepted, or this is just a note for a future?

**Comment by ObaniGemini** [comment]
You should maybe wait for other reviews, discuss about it on IRC.
Read !(CONTRIBUTING.md)[https://github.com/minetest/minetest/blob/master/CONTRIBUTING.md#code] to know more about the contributing process

**Comment by PilzAdam** [comment]
Why did you change `player:get_inventory():add_item()` to `player:add_item()`? This doesn't even work according to `lua_api.txt`.

**Comment by rubenwardy** [comment]
> Why did you change player:get_inventory():add_item() to player:add_item()? This doesn't even work according to lua_api.txt.

It looks like `player:add_item()` is a new function

**Comment by paramat** [comment]
> This way we can make, for example, overweight inventory effects.

Please could you explain more what this would be useful for? What is an overweight inventory effect?

**Comment by asl97** [comment]
@paramat i am guessing that it means slowing down the player movement (ie: set_physics_override) when the player carry too many stuff.

**Comment by Foghrye4** [comment]
@paramat asl97 guessed absolutely correctly. First i have idea to make a mod, which will limit player carrying capacity by weight of carried items, secondly i found out - there is not enought callbacks to do so, third - i made a nessesary callbacks and create this request.

**Comment by Foghrye4** [comment]
@paramat second option - detect whenever player take something radioactive, for example, enriched uranium fuel and add effects of radiation sickness. Or opposite.

**Comment by est31** [comment]
Can this be accomplished by not adding `add_item` and `get_item` methods to the player?

Otherwise, the idea is good, however you should do only one commit for this change, and add a meaningful commit description.

**Comment by Foghrye4** [comment]
@est31 commit? You mean - change only one file? Or do not change files after this request created?
There is no _get_item_ method in this pull request. Only _add_item_. 
About reason of this function to be added: of course i can hook up on InvRef _add_item_, but if i do so callback "on_inventory_add" will be launched on any inventory interactions, including node and detached inventories, unnecessary overlapping with they own callbacks. As well, its hard to extract player reference from this function.

**Comment by rubenwardy** [comment]
Each commit can update multiple files. There should only be one entry here ideally: https://github.com/minetest/minetest/pull/3962/commits

The merger can squash your commits, though

**Comment by PilzAdam** [comment]
This is missing documentation in `lua_api.txt`.

I would prefer if this would work without adding `player:add_item`, since it would require to re-write some mods. I wouldn't consider it bad that the callbacks may be overlapping; we have overlapping callbacks already, e.g. `minetest.register_on_dignode` and `nodedef.on_dig`.

**Comment by sofar** [comment]
- Lots of style issues (spaces after 'if', around '()' and '{}' not on the same lines as 'if' and 'else')
- Even if squashed, there's not a single commit that has an actual commit message that is sufficient.

**Comment by Foghrye4** [comment]
@PilzAdam  This could be done. But for my purposes i need player reference. And i have no idea how to extract player reverence from `InvRef::l_add_item(lua_State *L)`.
'lua_api.txt' was updated

**Comment by PilzAdam** [comment]
I don't like that the callbacks are only called if `player:add_item()` is used, but not if `player:get_inventory():add_item()` is called.

I think that the new `player:add_item()` function is only a workaround, because it's not simple to add it to the existing functions. I don't like this workaround.

:-1:

**Comment by Foghrye4** [comment]
@PilzAdam
I don't know a single mod, that should be updated for this implementation to work. It will work perfectly with 3d_armor, for example. Modders still may use whatever method they like - there is no holes, what players can exploit in currently presented mods.

**Comment by PilzAdam** [comment]
@Foghrye4 every mod that calls `player:get_inventory():add_item()` will cause the callbacks to be ignored for this action. Thus, mods which want to use your new callbacks can't rely on them.

**Comment by Foghrye4** [comment]
@PilzAdam 100% of them do so on `minetest.register_on_newplayer` function or by server commands. A first one is easily catched by `minetest.register_on_joinplayer` as in my example mod. A second one - is not my case.
Well, i do see your point here. Here is what i can do:
1. Add `ServerActiveObject` field in `InvRef` class.
2. Add `ServerActiveObject player` parameter in `InvRef::create(..)`
3. Add Lua API function `get_player()` in `InvRef`.
Should i use this method instead?

**Comment by rubenwardy** [comment]
Maybe some sort of InventoryChangeReceiver, which is implemented by Player? So

```cpp
class InventoryChangeReceiver
{
public:
    void on_move(...) = 0;
    void on_put(...) = 0;
};

class Inventory
{
public:
     InventoryChangeReceiver *rec;

    void do_move(...)
    {
        if (rec) rec->on_move(...);
    }
};
```

It could then be used to make node callbacks as well as player callbacks.
Don't implement this yet, just throwing the idea.
How are node metadata callbacks done currently?

**Comment by Foghrye4** [comment]
@rubenwardy 
> How are node metadata callbacks done currently?
Have no idea. As long as i can see - there is no such ones.

**Comment by rubenwardy** [comment]
from default:

```lua
minetest.register_node("default:chest", {
	on_metadata_inventory_move = function(pos, from_list, from_index,
			to_list, to_index, count, player)
		minetest.log("action", player:get_player_name() ..
			" moves stuff in chest at " .. minetest.pos_to_string(pos))
	end,
    on_metadata_inventory_put = function(pos, listname, index, stack, player)
		minetest.log("action", player:get_player_name() ..
			" moves " .. stack:get_name() ..
			" to chest at " .. minetest.pos_to_string(pos))
	end,
    on_metadata_inventory_take = function(pos, listname, index, stack, player)
		minetest.log("action", player:get_player_name() ..
			" takes " .. stack:get_name() ..
			" from chest at " .. minetest.pos_to_string(pos))
	end
})
```

**Comment by Foghrye4** [comment]
@rubenwardy here is a realisation of your idea. I will change some things (will remove Player parameter from ScriptApiPlayer functions and InvRef from Lua call, will update lua_api.txt).
It works perfectly fine with all three types of inventory: players, nodemeta and detached.

I would like to see your opinion and opinion of others members.

**Comment by Foghrye4** [comment]
There is 'minetest.clear_craft(recipe)' in this pull request as well here now. It was properly checked in all possible uses. This function allow modders remove previously registered recipes either by output or by input and type (including "fuel" and "cooking"). Plus, removed recipes no longer visible via crafting guides, because they eliminated from registry.

**Comment by est31** [comment]
@Foghrye4 can you open a separate PR for that, its unrelated to this change.

**Comment by Foghrye4** [comment]
@est31 As you wish.

**Pull request closed without merging**